src/app/@core/data/i18n.service.ts

```diff
@@ -0,0 +1,86 @@
+import { Injectable } from '@angular/core';
+import { TranslateService, LangChangeEvent } from '@ngx-translate/core';
+import { includes } from 'lodash';
+import { NbLayoutDirection, NbLayoutDirectionService } from '@nebular/theme';
+
+// TODO: mbed languages to avoid extra HTTP requests
+// import * as arEG from '../../../assets/translations/ar.json';
+// import * as enUS from '../../../assets/translations/en.json';
+
+const languageKey = 'language';
+
+/**
+ * Pass-through function to mark a string for translation extraction.
+ * Running `npm translations:extract` will include the given string by using this.
+ * @param {string} s The string to extract for translation.
+ * @return {string} The same string.
+ */
+export function extract(s: string) {
+  return s;
+}
+
+@Injectable()
+export class I18nService {
+
+  defaultLanguage: string;
+  supportedLanguages: string[];
+  directions = NbLayoutDirection;
+
+  constructor(private translateService: TranslateService, private directionService: NbLayoutDirectionService) {
+    // TODO: Embed languages to avoid extra HTTP requests
+    // translateService.setTranslation('en-US', enUS);
+    // translateService.setTranslation('ar-EG', arEG);
+  }
+
+  /**
+   * Initializes i18n for the application.
+   * Loads language from local storage if present, or sets default language.
+   * @param {!string} defaultLanguage The default language to use.
+   * @param {Array.<String>} supportedLanguages The list of supported languages.
+   */
+  init(defaultLanguage: string, supportedLanguages: string[]) {
+    this.defaultLanguage = defaultLanguage;
+    this.supportedLanguages = supportedLanguages;
+    this.language = '';
+
+    this.translateService.onLangChange
+      .subscribe((event: LangChangeEvent) => { localStorage.setItem(languageKey, event.lang); });
+  }
+
+  /**
+   * Sets the current language.
+   * Note: The current language is saved to the local storage.
+   * If no parameter is specified, the language is loaded from local storage (if present).
+   * @param {string} language The IETF language code to set.
+   */
+  set language(language: string) {
+    language = language || localStorage.getItem(languageKey) || this.translateService.getBrowserCultureLang();
+    let isSupportedLanguage = includes(this.supportedLanguages, language);
+
+    // If no exact match is found, search without the region
+    if (language && !isSupportedLanguage) {
+      language = language.split('-')[0];
+      language = this.supportedLanguages.find(supportedLanguage => supportedLanguage.startsWith(language)) || '';
+      isSupportedLanguage = Boolean(language);
+    }
+
+    // Fallback if language is not supported
+    if (!isSupportedLanguage) {
+      language = this.defaultLanguage;
+    }
+
+    language === 'ar-EG' ? this.directionService.setDirection(this.directions.RTL)
```

Comment:
Would there be ways to get this to scale better with other languages? Hardcoding `ar-EG` here and in other places seems risky

Reply:
The requirements was only for Arabic and English. Of course it could be handled by adding switch case